homeassistant/auth/providers/openid.py

```diff
@@ -0,0 +1,270 @@
+"""OpenId based authentication provider."""
+
+import logging
+import re
+from secrets import token_hex
+from typing import Any, Dict, Optional, cast
+
+from aiohttp import ClientResponseError
+from aiohttp.client import ClientResponse
+import voluptuous as vol
+from yarl import URL
+
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+from homeassistant.helpers.config_entry_oauth2_flow import (
+    AUTH_CALLBACK_PATH,
+    _encode_jwt,
+    async_register_view,
+)
+from homeassistant.helpers.network import get_url
+
+from . import AUTH_PROVIDER_SCHEMA, AUTH_PROVIDERS, AuthProvider, LoginFlow
+from ..models import Credentials, UserMeta
+
+REQUIREMENTS = ["python-jose==3.1.0"]
+
+_LOGGER = logging.getLogger(__name__)
+
+CONF_CLIENT_ID = "client_id"
+CONF_CLIENT_SECRET = "client_secret"
+CONF_CONFIGURATION = "configuration"
+CONF_EMAILS = "emails"
+
+CONFIG_SCHEMA = AUTH_PROVIDER_SCHEMA.extend(
+    {
+        vol.Required(CONF_CONFIGURATION): str,
+        vol.Required(CONF_CLIENT_ID): str,
+        vol.Required(CONF_CLIENT_SECRET): str,
+        vol.Required(CONF_EMAILS): [str],
+    },
+    extra=vol.PREVENT_EXTRA,
+)
+
+OPENID_CONFIGURATION_SCHEMA = vol.Schema(
+    {
+        vol.Required("issuer"): str,
+        vol.Required("jwks_uri"): str,
+        vol.Required("id_token_signing_alg_values_supported"): list,
+        vol.Optional("scopes_supported"): vol.Contains("openid"),
+        vol.Required("token_endpoint"): str,
+        vol.Required("authorization_endpoint"): str,
+        vol.Required("response_types_supported"): vol.Contains("code"),
+        vol.Optional(
+            "token_endpoint_auth_methods_supported", default=["client_secret_basic"]
+        ): vol.Contains("client_secret_post"),
+        vol.Optional(
+            "grant_types_supported", default=["authorization_code", "implicit"]
+        ): vol.Contains("authorization_code"),
+    },
+    extra=vol.ALLOW_EXTRA,
+)
+
+
+class InvalidAuthError(HomeAssistantError):
+    """Raised when submitting invalid authentication."""
+
+
+async def raise_for_status(response: ClientResponse) -> None:
+    """Raise exception on data failure with logging."""
+    if response.status >= 400:
+        standard = ClientResponseError(
+            response.request_info,
+            response.history,
+            code=response.status,
+            headers=response.headers,
+        )
+        data = await response.text()
+        _LOGGER.error("Request failed: %s", data)
+        raise InvalidAuthError(data) from standard
+
+
+WANTED_SCOPES = {"openid", "email", "profile"}
+
+
+@AUTH_PROVIDERS.register("openid")
+class OpenIdAuthProvider(AuthProvider):
+    """Example auth provider based on hardcoded usernames and passwords."""
+
+    DEFAULT_TITLE = "OpenId Connect"
```

Comment:
I'd like to point out that it's "OpenID Connect" instead of "OpenId Connect". The example docstring also shouldn't be there, but replaced with a proper one.

For my own use case I'd like to request the following features:

- Make it possible to disable requiring a verified e-mail address.
- Make it possible to use usernames instead of e-mail addresses.
- Make it possible to not have an e-mail address whitelist.

On the front-end, it would be nice if it did an actual redirect, or maybe a button to open the login pop-up, so that most browsers won't block the pop-up.

I also don't like how new users are admin by default, though that's also a problem with the CLI auth provider.

Reply:
Well the problem is email address is not considered the real persistent identifier. If the email is not validated, it is fully possible for somebody to register a new account with an invalid email address (somebody elses) on the service and suddenly be able to log into HASS. So i really don't think skipping validation check is okey.

The real identifier should be the subject identifier, which aught to be what is whitelisted. But that is somewhat of a random string, or at least unknown for the user. So i'm unsure how to get the user to figure that identifier out.

Reply:
The thing is, I have my own Keycloak server for SSO where I don't have e-mail addresses set for most users (so that's why I'd prefer it if I could set it to use e.g. `preferred_username` instead). It's also set up to only successfully authenticate the users that I want to give access to Home Assistant. Users also can't self-register in my Keycloak instance. I don't think that what I want should be default behavior though, just configurable in the YAML.

Reply:
Well openid is authentication not authorization. So we must have a way to whitelist.

I don't mind adding subject identifier whitelist though, and using preferred username only as the hass name.

You need to know the subject identifier though. Would be logged on failed logging.

Reply:
We could possibly whitelist roles claims.

Reply:
[Here](https://gist.github.com/wb9688/a8b0602bdd76d6716922dcc3664b62b0) is a _**quick and dirty**_ implementation of what I want.

Reply:
I've added some of that, but I will not accept not checking for verified email. We can add whitelist on subject, or perhaps roles.

I'll look at that button press thingy, i think we need chime in from frontend guys on that first.

Reply:
OK, but then don't check for verified e-mail if you whitelist with something other than e-mail. (I didn't look at your implementation yet, so maybe you already did that.)

I think the button press is necessary as:

1. Most browsers will block it without button press.
2. If the browser doesn't block it, you'll first have to close the pop-up to be able to select a different authentication method.

You should of course localize the "log in" text in the button (unlike I did, didn't I say it was quick and dirty?), and maybe move the button code to a different place.

Reply:
If subject matches, that is enough. Then the email scope is not even needed.

Reply:
Yes, I made that comment before looking at your actual implementation. Now I looked at it, I'm happy with your current implementation, though a role-based whitelist (not necessarily a list, just one allowed role is OK) would be nice.

My only actual issue with it now is the button press, though you indeed should probably ask the front-end guys.